pwiz_tools/Shared/PanoramaClient/PanoramaUtil.cs

```diff
@@ -675,37 +349,128 @@ private static string GetErrorMessage(FolderState state, string folderPath, stri
                 case FolderState.notpanorama:
                     stateError = string.Format(Resources.PanoramaUtil_VerifyFolder__0__is_not_a_Panorama_folder, folderPath);
                     break;
-                case FolderState.unknown:
-                    stateError = string.Format("Unrecognized error trying to get status for folder {0}.", folderPath);
-                    break;
             }
+            return stateError;
+        }
+    }
+
+    public class PanoramaServerException : Exception
+    {
+        public PanoramaServerException(string message) : base(message)
+        {
+        }
 
-            return AppendErrorAndUri(stateError, error, requestUri);
+        public PanoramaServerException(string message, Exception e) : base(message, e)
+        {
+        }
+    }
+
+    public class ErrorMessageBuilder
+    {
+        private readonly string _error;
+        private string _errorDetail;
+        private string _exceptionMessage;
+        private LabKeyError _labkeyError;
+        private Uri _uri;
+        private string _responseString;
+
+        public ErrorMessageBuilder (string error)
+        {
+            _error = error;
+        }
+        public ErrorMessageBuilder ErrorDetail(string errorDetail)
+        {
+            _errorDetail = errorDetail;
+            return this;
+        }
+        public ErrorMessageBuilder LabKeyError(LabKeyError labkeyError)
+        {
+            _labkeyError = labkeyError;
+            return this;
         }
 
-        private static string AppendErrorAndUri(string mainMessage, string error, Uri uri)
+        public ErrorMessageBuilder Exception(Exception ex)
         {
-            var message = mainMessage;
+            Exception(ex, null);
+            return this;
+        }
 
-            if (error != null || uri != null)
+        // If the given exception is a WebException, this method tries to parse the response as JSON and
+        // get a LabKey error if it was set. The method accepts a function to parse the response in the 
+        // WebException so that tests can pass in a custom function.
+        public ErrorMessageBuilder Exception(Exception ex, Func<WebException, LabKeyError> labkeyErrorGetter)

```

Comment:
It might be better if this method took an Exception and a LabKeyError.
It's unusual to pass a Func into a method when the caller could so easily have called the function themselves and passed in the result instead.